azurelinuxagent/ga/update.py

```diff
@@ -491,45 +496,94 @@ def _check_threads_running(self, all_thread_handlers):
                 logger.warn("{0} thread died, restarting".format(thread_handler.get_thread_name()))
                 thread_handler.start()
 
+    errors = 0
+
     def _try_update_goal_state(self, protocol):
         """
         Attempts to update the goal state and returns True on success or False on failure, sending telemetry events about the failures.
         """
+        max_errors_to_log = 3
+
         try:
-            max_errors_to_log = 3
+            #
+            # For Fast Track goal states we need to ensure that the tenant certificate is in the goal state.
+            #
+            # Some scenarios can produce inconsistent goal states. For example, during hibernation/resume, the Fabric goal state changes (the
+            # tenant certificate is re-generated when the VM is restarted) *without* the incarnation necessarily changing (e.g. if the incarnation
+            # is 1 before the hibernation; on resume the incarnation is set to 1 even though the goal state has a new certificate). If a Fast
+            # Track goal state comes after that, the extensions will need the new certificate.
+            #
+            # For new Fast Track goal states, we check the certificates and, if an inconsistency is detected,  re-fetch the entire goal state
+            # (update_goal_state(force_update=True). We re-fetch 2 times, one without waiting (to address scenarios like hibernation) and one with
+            # a delay (to address situations in which the HGAP and the WireServer are temporarily out of sync).
+            #
+            for attempt in range(3):
+                protocol.client.update_goal_state(force_update=attempt > 0, silent=self._update_goal_state_error_count >= max_errors_to_log, save_to_history=True)
+
+                goal_state = protocol.get_goal_state()
+                new_goal_state = self._goal_state is None or self._goal_state.extensions_goal_state.id != goal_state.extensions_goal_state.id
+
+                if not new_goal_state or goal_state.extensions_goal_state.source != GoalStateSource.FastTrack:
+                    break
```

Comment:
`Current code stops at retry 1, since previous fetch and current fetch same incarnation
`
This won't happen, since we update self._goal_state is updated only after all retries have been done

Reply:
Oh, Github PR UI deceived me :). self._goal_state is outside of loop, thanks for confirming, All good then